pkg/mimirpb/symbols_test.go

```diff
@@ -0,0 +1,123 @@
+package mimirpb
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/prometheus/prometheus/model/labels"
+	writev2 "github.com/prometheus/prometheus/prompb/io/prometheus/write/v2"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSymbolsTable(t *testing.T) {
+	// Ensure that the Mimir and Prometheus implementations share the same properties.
+	promImplBuilder := func() StringSymbolizer {
+		st := writev2.NewSymbolTable()
+		return &st
+	}
+	fastImplBuilder := func() StringSymbolizer {
+		st := NewFastSymbolsTable(baseSymbolsMapCapacity)
+		return st
+	}
+	impls := [](func() StringSymbolizer){promImplBuilder, fastImplBuilder}
+
+	for i, impl := range impls {
+		t.Run(fmt.Sprintf("impl %d", i), func(t *testing.T) {
+			s := impl()
+
+			require.Equal(t, []string{""}, s.Symbols(), "required empty reference does not exist")
+			require.Equal(t, uint32(0), s.Symbolize(""))
+			require.Equal(t, []string{""}, s.Symbols())
+
+			require.Equal(t, uint32(1), s.Symbolize("abc"))
+			require.Equal(t, []string{"", "abc"}, s.Symbols())
+
+			require.Equal(t, uint32(2), s.Symbolize("__name__"))
+			require.Equal(t, []string{"", "abc", "__name__"}, s.Symbols())
+
+			require.Equal(t, uint32(3), s.Symbolize("foo"))
+			require.Equal(t, []string{"", "abc", "__name__", "foo"}, s.Symbols())
+
+			s.Reset()
+			require.Equal(t, []string{""}, s.Symbols(), "required empty reference does not exist")
+			require.Equal(t, uint32(0), s.Symbolize(""))
+
+			require.Equal(t, uint32(1), s.Symbolize("__name__"))
+			require.Equal(t, []string{"", "__name__"}, s.Symbols())
+
+			require.Equal(t, uint32(2), s.Symbolize("abc"))
+			require.Equal(t, []string{"", "__name__", "abc"}, s.Symbols())
+
+			ls := labels.FromStrings("__name__", "qwer", "zxcv", "1234")
+			encoded := make([]uint32, len(ls)*2)
+			for i, l := range ls {
+				encoded[(i * 2)] = s.Symbolize(l.Name)
+				encoded[(i*2)+1] = s.Symbolize(l.Value)
+			}
+			require.Equal(t, []uint32{1, 3, 4, 5}, encoded)
+			decoded, _ := desymbolizeLabelsDirect(encoded, s.Symbols())
+			require.Equal(t, ls, FromLabelAdaptersToLabels(decoded))
+		})
+	}
+}
+
+func desymbolizeLabelsDirect(labelRefs []uint32, symbols []string) ([]LabelAdapter, string) {
+	name := ""
+	las := make([]LabelAdapter, len(labelRefs)/2)
+	for i := 0; i < len(labelRefs); i += 2 {
+		las[i/2].Name = symbols[labelRefs[i]]
+		las[i/2].Value = symbols[labelRefs[i+1]]
+		if las[i/2].Name == labels.MetricName {
+			name = las[i/2].Value
+		}
+	}
+	return las, name
+}
+
+func BenchmarkSymbolizer(b *testing.B) {
+	b.Run("prom symbolizer: 10k labels unique values", func(b *testing.B) {
```

Comment:
Individual sends from distributors to ingesters typically have 7 series, each with say 20 labels, so 280 strings total. That's the sort of ball-park we should be benchmarking.

Reply:
Agree on few series per per-partition write request. Let's say `max_samples_per_send` is 2000 and requests are full. Then we have per-partition write requests within this range:
- A tenant with shard size = 1: 2000 symbols in the per-partition write request
- A tenant with shard size = all ingesters: with 100 partitions that's 200 samples per per-partition write request, with 300 partitions (which is pretty large scale) that's 7, the number Bryan mentioned

Reply:
I've updated this benchmark to be more realistic (a case for both bryan's and marco's scenarios). It's also not coupled to the implementation so much anymore, thus easier to use with `benchstat` and iterate on.